haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs

```diff
@@ -177,82 +194,72 @@ declsFromRenamed ctx (hsGroup, _, _, _) =
         (instDefs, instChanges) = second M.unions . unzip . mapMaybe instDecls
                                 . hs_instds
                                 $ hsGroup
-        varBindDefs = concatMap declsFromForall . universeBi $ hsGroup
+        explicitVarBindDefs = map mkDeclName explicitNames
         -- TODO(robinpalotai): emit DeclMods for type signatures, so the later
         --   extracted decls can carry them.
-    in (concat [defs, instDefs, varBindDefs], instChanges)
+    in (concat [defs, instDefs, explicitVarBindDefs, implicitVarBindDefs], instChanges)
   where
-    -- Getting the location of implicit variable bindings is tricky, as the
-    -- HsTyVarBndr's location is the whole span of the scope, at least for the
-    -- HsForAllTy. The actual location is harvested from the usage
-    -- locations, ideally from the one which occurs first by location.
-    -- TODO(robinpalotai): it seems a bit dodgy that the first usage emits
-    -- both the declaration and the usage. The decl should go to some virtual
-    -- location / figment.
-    declsFromForall :: HsType Name -> [DeclAndAlt]
-    declsFromForall (HsForAllTy explicitness _ binders context ty) =
-        -- TODO(robinpalotai): a kind signature will include the kind in the
-        --   loc, need heuristic to strip it off (also for binds from data
-        --   decls).
-        case explicitness of
-            Explicit -> map (mkDecl . unLoc) . hsq_tvs $ binders
-              where
-                mkDecl binder =
-                    nameDeclAlt ctx (boundTyVarName binder) Nothing
-                                typeStringyType
-            Implicit  -> implicitCase
-            Qualified -> implicitCase
-                -- ^ Quoting Note [HsForAllTy tyvar binders]:
-                --
-                -- "Qualified currently behaves exactly as Implicit, but it is
-                -- deprecated to use it for implicit quantification."
+    mkDeclName n = nameDeclAlt ctx n Nothing typeStringyType
+    explicitNames = concatMap namesFromForall . universeBi $ hsGroup
+    implicitVarBindDefs =
+      let allUsages = concatMap namesFromTyVar . universeBi $ hsGroup
```

Comment:
nit: please use 4-space indents (sorry, will have to add to Readme). `where` is ok to have 2-space indent (and then next line again 2 to get to total 4). If you don't want to bother, can leave like this and I'll fix when merging.